[core] Feature: high resolution scrolling and extended scroll reports

Alabastard-64 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
### **Current version of this PR requires that PR #18218 be merged first**
From feature request #17585 and partially based on @fauxpark 's test code, this adds an option to enable high resolution scrolling which allows some apps to have sub line scrolling (down to pixel resolution if desired).  

This requires that scrolling sensitivity is adjusted based on what `MOUSE_SCROLL_MULTIPLIER` is set to.  Currently most mainstream applications such as firefox, chrome and discord support the feature and most other applications that let the OS handle scroll input will behave the same as they did before (as long as scroll sensitivity is adjusted properly).  Unfortunately there are a few applications that do not handle this well, and the scroll sensitivity adjustment to compensate for hires scrolling can make those apps oversensitive to scroll inputs, this can be mitigated by using a precision toggle when scrolling in these apps or just adapting to the increased sensitivity.

Drag scroll mode included with the pointing modes feature will automatically adjust scroll sensitivity if it detects that hires scrolling has been enabled (which has to be done by the OS by sending a set_report feature request to the keyboard).

Also the option to extend scroll reports to 16 Bit has also been included which I have actually found is pretty useful and reduces code size a bit with pointing modes enabled (PR #18218 ).

#### **Currently only Atmega U4 chip code has been tested and is working with both `MOUSE_SHARED_EP = yes/no`**
USB Code has been ported to `vusb.c` and `ChibiOS/usb_main.c` to allow for U2 and ARM chip support but is currently untested so I will leave this in draft until I can confirm from someone it is working.

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* #17585 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

freznel10 commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -610,6 +616,14 @@ static void set_led_transfer_cb(USBDriver *usbp) {
+#ifdef MOUSE_SCROLL_HIRES_ENABLE
+static void set_multiplier_cb(USBDriver *usbp) {
+    if (usbp->setup[6] == 2 && set_report_buf[0] == REPORT_ID_MULTIPLIER) {
+        resolution_multiplier = set_report_buf[1]
```
I think this is missing a ";"

Alabastard-64 replied:
Yup,
Thanks for testing this!
Will add in the next update I did not even test if the vusb or ChibiOS code compiled...

freznel10 commented:
I can confirm this works really well on an RP2040

BanchouBoo commented on tmk_core/protocol/report.h:
```diff
@@ -207,8 +208,23 @@ typedef int16_t mouse_xy_report_t;
+#ifdef MOUSE_SCROLL_HIRES_ENABLE
+#    ifndef MOUSE_SCROLL_MULTIPLIER
+#        define MOUSE_SCROLL_MULTIPLIER 120
+#    elif (MOUSE_SCROLL_MULTIPLIER >= 120 || MOUSE_SCROLL_MULTIPLIER < 1)
```
This checks of `MOUSE_SCROLL_MULTIPLIER` check if it's greater or *equal to* 120, but the default value is 120 and the error message says that 120 is a valid value.

Alabastard-64 replied:
Yeah that is a mistake on my part thanks for catching that will put it in the next update.

BanchouBoo commented:
Testing this with `ploopyco/trackball` with a multiplier of 120.

On my scroll wheel, when going from a standstill I have to scroll a *lot* in one direction before it starts scrolling at all, at which point it will first scroll 30 pixels in that direction after which it will start scrolling one pixel at a time after that until I stop scrolling, after which point it resets back to that initial state.

The same seems to be true for drag scroll as well (tested from manually mapping x/y to h/v in `pointing_device_task_user`, I tried with `PM_MO(PM_DRG)` as well but that didn't do anything, my mouse pointer remained active), but it's a bit harder to get it to start scrolling smoothly consistently, but I don't think that's because of any unique problem occuring with drag scroll so much as it's just harder to observe when compared to the scroll wheel which only updates after moved a certain distance rather than with every movement.

As an aside to this bug report, how would one set it up so that it doesn't effect their scroll wheel while still allowing drag scroll to use it?

Alabastard-64 commented:
> On my scroll wheel, when going from a standstill I have to scroll a _lot_ in one direction before it starts scrolling at all, at which point it will first scroll 30 pixels in that direction after which it will start scrolling one pixel at a time after that until I stop scrolling, after which point it resets back to that initial state.

This is actually normal behaviour believe it or not when you do not adjust the scroll sensitivity when high resolution is active.  I need to better explain this in the documentation (_I think I will need to expand it to a whole section_).

### What is happening:
___NOTE: The following is based on the Windows Vista+ implementation need to re-track down the Linux implementation (only ) but it was originally based off of copying the Windows implementation___

When hi-res scrolling is active the OS now treats all v/h axis reports differently so that the following calculation is performed:  

$$scroll_{hires} = \frac{scroll \times 120}{MOUSE\\_SCROLL\\_MULTIPLIER}$$

and that is treated as the new input from that device.  However, $scroll_{hi-res}$ is intended to be the number of pixels scrolled rather than number of lines so the OS (_at least in windows_) will do something else for applications that do not support per pixel scrolling (_which includes windows explorer oddly enough_):   

$$lines\\_scrolled = \frac{scroll_{hires}}{120}$$

Now since in your case you had resolution multiplier at 120 you would need a large enough movement to have an output >120 in order to move a single line in those applications resulting in the jerky clunky movement you observed.  Usually to compensate for this the output of the v/h will need to be increased in some way.

Most implementations of high res scrolling I've seen do this with a couple of approaches high sensitivity or CPI setting on the mouse/trackball and use a large divisor when hires mode is not active.

For example pointing modes I added the following calculation to drag scroll:  

```c
    switch (pointing_mode.id) {
        // drag scroll mode (sets mouse axes to mouse_report h & v with divisor)
        case PM_DRAG:
#    ifdef MOUSE_SCROLL_HIRES_ENABLE
           {
                uint8_t drag_multiplier = MAX(MOUSE_SCROLL_MULTIPLIER / pointing_mode.divisor, 1);
                if (resolution_multiplier & 1 << 0) pointing_mode.y *= drag_multiplier;
                if (resolution_multiplier & 1 << 2) pointing_mode.x *= drag_multiplier;
                if (resolution_multiplier) pointing_mode.divisor = 1;
           }
#    endif
```
Note that with the above you will not get the full benefit of the hires scrolling in the above case unless the `POINTING_DRAG_DIVISOR >= MOUSE_SCROLL_MULTIPLIER` but you will get some benefit as long as `POINTING_DRAG_DIVISOR > 1`.

Also Note `resolution_multiplier`  is a global variable set by the OS during USB initialization.
***I will add some functions/macros to simplify using this in peoples code***

> I used https://cpstest.org/scroll-test.php to see how much it was scrolling. For reference, when I don't have this feature enabled, my scroll values on that site are in increments of 60. I'm also on Linux, unsure if that's relevant.

Thanks this is helpful! seems that hires scrolling is not working properly for you as I am able to get increments of 1 when hires is enabled using pointing_modes drag scroll will need to review linux implementation (_I hate how poorly documented this is on all OS's_).

> The same seems to be true for drag scroll as well (tested from manually mapping x/y to h/v in `pointing_device_task_user`, I tried with `PM_MO(PM_DRG)` as well but that didn't do anything, my mouse pointer remained active), 

This seems like pointing\_modes is not working properly for you... pointer should be stationary when a key with PM_MO(PM_DRG) pressed.  Did you `#define POINTING_DEVICE_MODES_ENABLE` somewhere in  one of your `config.h` file's (_currently DD is unsupported for pointing\_modes _)?

> but it's a bit harder to get it to start scrolling smoothly consistently, but I don't think that's because of any unique problem occuring with drag scroll so much as it's just harder to observe when compared to the scroll wheel which only updates after moved a certain distance rather than with every movement.

I'm not sure it is fully working for you, so  thanks for the report! I'll try and dig into what is going on here and see if it is due to differences between Linux and Windows (which I assumed would work the same as the Linux implementation was built to mimic the windows originally, _No idea about MacOS..._) or something else is going on. as the `ploopyco/trackball` uses a atmegaU4 which what I'm currently using so it should work.

> As an aside to this bug report, how would one set it up so that it doesn't effect their scroll wheel while still allowing drag scroll to use it?

Interesting idea!  not sure if it would work as it is unclear when the OS will request hires scrolling settings from the device (the OS sets these settings during USB initialization and then checks them ) but it could be worth testing!  setting the resolution_multiplier to zero should work but I have no idea when the best time to re-enable it would be (you would at least need to wait until after the report is sent).

BanchouBoo commented:
> This seems like pointing_modes is not working properly for you... pointer should be stationary when a key with PM_MO(PM_DRG) pressed. Did you #define POINTING_DEVICE_MODES_ENABLE somewhere in one of your config.h file's (_currently DD is unsupported for pointing_modes _)?

Ah, I had `_ENABLED` rather than `_ENABLE`, pointing modes work now.

> Thanks this is helpful! seems that hires scrolling is not working properly for you as I am able to get increments of 1 when hires is enabled using pointing_modes drag scroll will need to review linux implementation (I hate how poorly documented this is on all OS's).

Without changing the divisor, 30 is still the lowest value I'm getting, then it scrolls in increments of 15.

> Interesting idea! not sure if it would work as it is unclear when the OS will request hires scrolling settings from the device (the OS sets these settings during USB initialization and then checks them ) but it could be worth testing! setting the resolution_multiplier to zero should work but I have no idea when the best time to re-enable it would be (you would at least need to wait until after the report is sent).

Would it be possible to have it set up so that instead of `MOUSE_SCROLL_MULTIPLIER` being a macro at compile time, it's a modifiable runtime value?

Alternatively, would it be viable to use `pointing_device_task_user` to mathematically undo whatever is going on to apply `resolution_multiplier`?

Alabastard-64 commented:
> Without changing the divisor, 30 is still the lowest value I'm getting, then it scrolls in increments of 15.

Okay this sounds about right now, also there was a bit of a bug in horizontal scrolling that I just fixed will push update soon. you would need to set the divisor to >=120 to get full pixel perfect scrolling but you would likely need to increase CPI to compensate for the low sensitivity.

>Would it be possible to have it set up so that instead of MOUSE_SCROLL_MULTIPLIER being a macro at compile time, it's a modifiable runtime value?

Unfortunately not in any reasonable way as this value is given as part of the USB HID Descriptor during initialization so it would be quite challenging to make this dynamic and even then you would need to re initialize the USB device driver every time the value changes.

> Alternatively, would it be viable to use pointing_device_task_user to mathematically undo whatever is going on to apply resolution_multiplier?

Yes you could, I would recommend using the pointing\_mode divisors\adding multipliers as well for drag scroll but either works.

However there is a way I could simply allow for the disabling the resolution multiplier temporarily until after the next mouse_report is sent (_It would be difficult to know when to re-enable otherwise_).

Alabastard-64 commented:
New feature in the new version:
you can now call the `hires_scroll_disable_next(hires_axis_t axis)` function with axis values `HIRES_V`, `HIRES_H`, or `HIRES_BOTH` to disable hires scrolling on v/h/both axes the next mouse report (will be re-enabled once mouse_report is sent).

Alabastard-64 commented:
> Getting some error now when I try to compile
> 
> ```
> quantum/pointing_device/pointing_device_modes.c:366:13: error: a label can only be part of a statement and a declaration is not a statement
>   366 |             uint8_t cur_divisor = pointing_mode.divisor;
>       |             ^~~~~~~
> quantum/pointing_device/pointing_device_modes.c:367:13: error: expected expression before 'uint8_t'
>   367 |             uint8_t drag_multiplier = MAX(MOUSE_SCROLL_MULTIPLIER / cur_divisor, 1);
>       |             ^~~~~~~
> quantum/pointing_device/pointing_device_modes.c:369:36: error: 'drag_multiplier' undeclared (first use in this function)
>   369 |                 pointing_mode.x *= drag_multiplier;
>       |                                    ^~~~~~~~~~~~~~~
> quantum/pointing_device/pointing_device_modes.c:369:36: note: each undeclared identifier is reported only once for each function it appears in
>  [ERRORS]
>  |
>  |
>  |
> make[1]: *** [builddefs/common_rules.mk:360: .build/obj_ploopyco_trackball_rev1_005_boo/quantum/pointing_device/pointing_device_modes.o] Error 1
> Make finished with errors
> ```

My bad I pushed before I added the scope for those variables and forgot to push the update.

Also there were a number of errors in the latest build I did not realize that I was not properly testing it for the last couple versions will update with a fixed version soon currently horizontal scrolling is broken again.

Alabastard-64 commented:
>> Without changing the divisor, 30 is still the lowest value I'm getting, then it scrolls in increments of 15.

>~~Okay this sounds about right now, also there was a bit of a bug in horizontal scrolling that I just fixed will push update soon. you would need to set the divisor to >=120 to get full pixel perfect scrolling but you would likely need to increase CPI to compensate for the low sensitivity.~~

I was wrong if this is happening it means the hires scrolling is not working properly.  After testing properly with a track pad and properly enabling hires scrolling, I was able to get 1-2 pixel resolution on that test, even with a smaller divisor (***8***).  I'll update the description that it is still currently not tested/working on Linux.

@BanchouBoo 
What Distro and version of Linux are you running?
hi-res scrolling support is pretty new on Linux (_1 year at most_)

BanchouBoo commented:
I'm on Void Linux, kernel version 6.1.3

Alabastard-64 commented:
> Void Linux, kernel version 6.1.3

Okay well after digging through the HID_input.c code in the 6.1.3 kernel it seems Hi-res scrolling should be supported now I guess I just need to dig into why it seems to not be working 100% for you.

Would you be willing to test a few things for me?

1. What do you get for pixel counts when high res scrolling is off?

2. Did you also enable `MOUSE_SCROLL_EXTENDED_REPORT`?

3. You mentioned that you are getting the slow down from the "/120" that seems to be happening but not the per pixel scrolling right? (Now that I've tested them side by side I realize that it is pretty obvious when it is working properly but subtle)

BanchouBoo commented:
> What do you get for pixel counts when high res scrolling is off?

I get increments of 60 for both my scroll wheel and drag scroll when hi-res scrolling is turned off

> Did you also enable MOUSE_SCROLL_EXTENDED_REPORT?

Yes

> You mentioned that you are getting the slow down from the "/120" that seems to be happening but not the per pixel scrolling right? (Now that I've tested them side by side I realize that it is pretty obvious when it is working properly but subtle)

With hi-res scrolling turned on, scrolling doesn't do anything for awhile and then jumps 30 units (always exactly 30) all at once, for both the wheel and drag scroll, after that point if I keep scrolling I get what seems to be proper precision if I continue scrolling, and if I stop then I have to break through that barrier again and get that 30 unit jump again, though drag scroll in these conditions still doesn't feel as smooth as it should, but that might be that it's resetting to whatever that base state is more often than the wheel for whatever reason, but the wheel absolutely scrolls pixel by pixel per each scroll tick once I break that barrier.

With the `PM_DRAG` pointing mode with the default divisor, instead of increments of 1 it's increments of 15. With a divisor of 120 or by mapping the x/y values to h/v in `pointing_device_task_user` instead of using pointing modes, it's increments of 1.

So, once past whatever that barrier is, I believe it's behaving correctly, it's just the barrier itself that's the problem.

Unsure if it's potentially relevant but probably worth noting, I'm using the `pointing_device_task_user` callback in my firmware. It's not doing anything to the scrolling values, I'm only using it to apply pointer acceleration, but once the PR is in a buildable state again I can try removing any extra behavior that isn't directly relevant to hi-res scrolling and see if that changes anything.

drashna commented:
Looks like if you use `MOUSE_SHARED_EP = no` to disable the shared EP:

```
Compiling: tmk_core/protocol/lufa/lufa.c                                                           tmk_core/protocol/lufa/lufa.c: In function 'EVENT_USB_Device_ControlRequest':
tmk_core/protocol/lufa/lufa.c:463:26: error: 'MOUSE_INTERFACE' undeclared (first use in this function)
                     case MOUSE_INTERFACE:
                          ^
tmk_core/protocol/lufa/lufa.c:463:26: note: each undeclared identifier is reported only once for each function it appears in
```

Alabastard-64 commented:
> Looks like if you use `MOUSE_SHARED_EP = no` to disable the shared EP:
> 
> ```
> Compiling: tmk_core/protocol/lufa/lufa.c                                                           tmk_core/protocol/lufa/lufa.c: In function 'EVENT_USB_Device_ControlRequest':
> tmk_core/protocol/lufa/lufa.c:463:26: error: 'MOUSE_INTERFACE' undeclared (first use in this function)
>                      case MOUSE_INTERFACE:
>                           ^
> tmk_core/protocol/lufa/lufa.c:463:26: note: each undeclared identifier is reported only once for each function it appears in
> ```

Huh I thought I tested that and fixed that error.  Guess it must have crept back in from a recent commit, thanks will fix it in the upcoming update just need to figure out what is happening with horizontal scroll... seems to be working but is cause wildly erratic behaviour when you are on any element that responds to either scroll input.

Alabastard-64 commented:
Well I have figured out the issue with horizontal scrolling for some reason it is not being set properly by the OS (at least in windows) not fixed in current version but I should have fixed all of the other issues.

BanchouBoo commented:
> Unsure if it's potentially relevant but probably worth noting, I'm using the pointing_device_task_user callback in my firmware. It's not doing anything to the scrolling values, I'm only using it to apply pointer acceleration, but once the PR is in a buildable state again I can try removing any extra behavior that isn't directly relevant to hi-res scrolling and see if that changes anything.

Tested to be sure, doesn't change anything. Removed everything except the keymap and the necessary portions of `config.h`.

Alabastard-64 commented:
Okay now would be a good time to test again and see if it is working for everyone alright I have fixed up a lot in this last update and both vertical and horizontal scrolling are working properly.

TODO:  
- [ ] Linting
- [ ] Add auto adjust to scroll mouskeys
- [ ] Confirm and test that adjusting multiplier works as intended
- [ ] Confirm new HID Descriptor still works on Linux (could use @BanchouBoo 's help here)
- [ ] Update and cleanup docs

BanchouBoo commented:
> Confirm new HID Descriptor still works on Linux (could use @BanchouBoo 's help here)

Drag scroll with the default divisor is smoother, but beyond that everything is the same, that 30 unit barrier still exists, so while it hasn't gotten any worse, it's still not fixed.

Alabastard-64 commented:
> > Unsure if it's potentially relevant but probably worth noting, I'm using the pointing_device_task_user callback in my firmware. It's not doing anything to the scrolling values, I'm only using it to apply pointer acceleration, but once the PR is in a buildable state again I can try removing any extra behavior that isn't directly relevant to hi-res scrolling and see if that changes anything.
> 
> Tested to be sure, doesn't change anything. Removed everything except the keymap and the necessary portions of `config.h`.

Sorry to bug you about this but I just want to confirm so I can understand what is happening when you use both `MOUSE_SCROLL_HIRES_ENABLE` and `POINTING_DEVICE_MODES_ENABLE`:
1. The older version you tested and the latest behave the same?
2. You have step sizes of 60 pixels/s in scrolling when using only `POINTING_DEVICE_MODES_ENABLE`?
3. When enabling both `MOUSE_SCROLL_HIRES_ENABLE` and `POINTING_DEVICE_MODES_ENABLE` with the default divisor you have step sizes of 30 pixels/s in scrolling?
4. Increasing the `POINTING_DRAG_DIVISOR` above the default of 8 decreases the step sizes in scrolling you get but never goes below 15?
5. Do you have `MOUSE_SHARED_EP = no` in `rules.mk` or `usb.shared_endpoint.mouse : false` in `info.json`?(_i.e. is your mouse report shared?_)
6. What is the polling rate you get when testing on this page [here](https://devicetests.com/mouse-rate-test)? (_I've noticed some latency effects on the scrolling rate at least in windows, should be ~800-1000 Hz by qmk defaults_)

Sorry for all the questions, and thanks for helping with this!

BanchouBoo commented:
> The older version you tested and the latest behave the same?
> Increasing the POINTING_DRAG_DIVISOR above the default of 8 decreases the step sizes in scrolling you get but never goes below 15?

Drag scroll with the default divisor is smoother, no longer in increments of 15, but everything else seems the same.

> You have step sizes of 60 pixels/s in scrolling when using only POINTING_DEVICE_MODES_ENABLE?

Correct, all scrolling without hi-res scrolling enabled are in increments of 60.

> When enabling both MOUSE_SCROLL_HIRES_ENABLE and POINTING_DEVICE_MODES_ENABLE with the default divisor you have step sizes of 30 pixels/s in scrolling?

Not quite. I scroll in one direction, for a bit nothing happens, then it scrolls 30 units at once, then if I keep scrolling after that it is seemingly pixel-precise beyond that point. If I stop scrolling it happens all over again the next time I scroll.

> Do you have MOUSE_SHARED_EP = no in rules.mk or usb.shared_endpoint.mouse : false in info.json?(i.e. is your mouse report shared?)

My `rules.mk` file is empty and the `info.json` for `ploopyco/trackball` does not contain anything related to `shared_endpoint`. The default `rules.mk` also does not contain an entry for `MOUSE_SHARED_EP` so I'm unsure what value it would be set to in my mouse. If you'd like me to try it with a particular value (or values), let me know.

> What is the polling rate you get when testing on this page here? (I've noticed some latency effects on the scrolling rate at least in windows, should be ~800-1000 Hz by qmk defaults)

Max value I hit was 910, and some additional research indicates that the hardware for the trackball can do 1000.

> Sorry for all the questions, and thanks for helping with this!

Don't be sorry! I want to get this working as much as you do, smooth drag scroll is the only pointer feature I have to rely on a software for currently, so if I can get it working in firmware that removes the need to rely on software at all and makes all my customization portable between computers.

Alabastard-64 commented:
Ah okay so I was a bit confused about the issue! Good thing I asked, but I think I understand now:
1. Hires scrolling does seem to be "working" at least partially but the behaviour is not quite right.
2. The lull before the first scroll event is confusing, I thought that you were talking about the how some apps will only move every MOUSE_SCROLL_MULTIPLIER units of scroll as they do not support sub line scroll movement.
   - but if this is happening even with the scrolling tester then something is very wrong and does not match the windows vista+ behaviour at all...

What software are you using to enable hires scrolling after the fact?  is it open source?  maybe I could see what they are doing to avoid this. (_assuming the issue is not also happening with the software_)

BanchouBoo commented:
It's a feature in [libinput](https://freedesktop.org/wiki/Software/libinput). Works perfectly through software, but only applies to drag scroll.

Not specifically useful or anything but this is the script I use to toggle it on and off.

```sh
#!/bin/sh

device="PloopyCo Trackball Mouse"
current_scroll_button=$(xinput --list-props "$device" | grep 'libinput Button Scrolling Button' | head -n1 | cut -f3)
scroll_button=8
current_value=$(xinput --list-props "$device" | grep 'libinput Scroll Method Enabled' | head -n1 | cut -d' ' -f7)
if [ "$current_scroll_button" -ne "$scroll_button" ]; then
    xinput --set-prop "$device" 'libinput Scroll Method Enabled' 0 0 1
    xinput --set-prop "$device" 'libinput Button Scrolling Button' "$scroll_button"
else
    new_value=$((current_value == 0))
    xinput --set-prop "$device" 'libinput Scroll Method Enabled' 0 0 "$new_value"
    xinput --set-prop "$device" 'libinput Button Scrolling Button' "$scroll_button"
fi
```

MatthiasGrandl commented:
Hi! I used to run this PR. Just updated and now I am getting this error while compiling:

```
tmk_core/protocol/chibios/usb_main.c:960: error: unterminated #ifdef
  960 | #ifdef MOUSE_ENABLE
      | 
tmk_core/protocol/chibios/usb_main.c:718: error: unterminated #if
  718 | #if defined(MOUSE_ENABLE) && !defined(MOUSE_SHARED_EP)
      | 
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
  716 |                             case SHARED_INTERFACE:
      |                             ^~~~
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:624:26: warning: unused variable 'dp' [-Wunused-variable]
  624 |     const USBDescriptor *dp;
      |                          ^~
tmk_core/protocol/chibios/usb_main.c: At top level:
tmk_core/protocol/chibios/usb_main.c:73:13: warning: 'keyboard_idle_timer_cb' declared 'static' but never defined [-Wunused-function]
   73 | static void keyboard_idle_timer_cb(struct ch_virtual_timer *, void *arg);
      |             ^~~~~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:623:13: warning: 'usb_request_hook_cb' defined but not used [-Wunused-function]
  623 | static bool usb_request_hook_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:615:13: warning: 'set_multiplier_cb' defined but not used [-Wunused-function]
  615 | static void set_multiplier_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:603:13: warning: 'set_led_transfer_cb' defined but not used [-Wunused-function]
  603 | static void set_led_transfer_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:580:17: warning: 'get_hword' defined but not used [-Wunused-function]
  580 | static uint16_t get_hword(uint8_t *p) {
      |                 ^~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:503:13: warning: 'usb_event_cb' defined but not used [-Wunused-function]
  503 | static void usb_event_cb(USBDriver *usbp, usbevent_t event) {
      |             ^~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:120:29: warning: 'usb_get_descriptor_cb' defined but not used [-Wunused-function]
  120 | static const USBDescriptor *usb_get_descriptor_cb(USBDriver *usbp, uint8_t dtype, uint8_t dindex, uint16_t wIndex) {
      |                             ^~~~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:71:24: warning: 'keyboard_idle_timer' defined but not used [-Wunused-variable]
   71 | static virtual_timer_t keyboard_idle_timer;
      |                        ^~~~~~~~~~~~~~~~~~~
 [ERRORStmk_core/protocol/chibios/usb_main.c:960: error: unterminated #ifdef
  960 | #ifdef MOUSE_ENABLE
      | 
tmk_core/protocol/chibios/usb_main.c:718: error: unterminated #if
  718 | #if defined(MOUSE_ENABLE) && !defined(MOUSE_SHARED_EP)
      | 
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
  716 |                             case SHARED_INTERFACE:
      |                             ^~~~
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:716:29: error: expected declaration or statement at end of input
tmk_core/protocol/chibios/usb_main.c:624:26: warning: unused variable 'dp' [-Wunused-variable]
  624 |     const USBDescriptor *dp;
      |                          ^~
tmk_core/protocol/chibios/usb_main.c: At top level:
tmk_core/protocol/chibios/usb_main.c:73:13: warning: 'keyboard_idle_timer_cb' declared 'static' but never defined [-Wunused-function]
   73 | static void keyboard_idle_timer_cb(struct ch_virtual_timer *, void *arg);
      |             ^~~~~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:623:13: warning: 'usb_request_hook_cb' defined but not used [-Wunused-function]
  623 | static bool usb_request_hook_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:615:13: warning: 'set_multiplier_cb' defined but not used [-Wunused-function]
  615 | static void set_multiplier_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:603:13: warning: 'set_led_transfer_cb' defined but not used [-Wunused-function]
  603 | static void set_led_transfer_cb(USBDriver *usbp) {
      |             ^~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:580:17: warning: 'get_hword' defined but not used [-Wunused-function]
  580 | static uint16_t get_hword(uint8_t *p) {
      |                 ^~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:503:13: warning: 'usb_event_cb' defined but not used [-Wunused-function]
  503 | static void usb_event_cb(USBDriver *usbp, usbevent_t event) {
      |             ^~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:120:29: warning: 'usb_get_descriptor_cb' defined but not used [-Wunused-function]
  120 | static const USBDescriptor *usb_get_descriptor_cb(USBDriver *usbp, uint8_t dtype, uint8_t dindex, uint16_t wIndex) {
      |                             ^~~~~~~~~~~~~~~~~~~~~
tmk_core/protocol/chibios/usb_main.c:71:24: warning: 'keyboard_idle_timer' defined but not used [-Wunused-variable]
   71 | static virtual_timer_t keyboard_idle_timer;
      |                        ^~~~~~~~~~~~~~~~~~~
 [ERRORS]
```

This is using the unmodified branch from this PR.

Alabastard-64 commented:
> Hi! I used to run this PR. Just updated and now I am getting this error while compiling:
> ...
> This is using the unmodified branch from this PR.

Really sorry about that added that error when I rebased and currently I've not been testing the ChibiOS code for compiler errors.

Let me know if that fixes your issue

freznel10 commented on tmk_core/protocol/chibios/usb_main.c:
```diff
@@ -683,20 +715,35 @@ static bool usb_request_hook_cb(USBDriver *usbp) {
                             case SHARED_INTERFACE:
 #endif
+#if defined(MOUSE_ENABLE) && !defined(MOUSE_SHARED_EP)
+                            case MOUSE_INTERFACE:
```
this needs an #endif ?

After I added that from the latest commit, the firmware compiles (using an RP2040)

Alabastard-64 replied:
yup, thanks will add to upcoming commit!
Really need to set up a dummy RP2040 build to test compiles or maybe I could just try compiling your board, Scylla something right?

freznel10 reviewed (commented):
Just linting

Alabastard-64 commented:
Currently working on updating pointing_device_modes which will be pushed to this PR as soon as I finish (which should address the all the conflicts)

So far getting hires scrolling to work without issue has been a bit of a nightmare as it seems to be handled inconsistently even by the same OS.
1. It works fine on a Windows 10 desktop(home machine) with no issues.
2. However doesn't work at all on a Windows 10 laptop(work machine) even though the laptop has a touch pad that uses hires scrolling!(may need to see if copying aspects of the touch pads HID report structure may help here).
   - This may be due to the apps that I am limited to on this machine (ms edge browser  etc.) or some settings from my companies IS/IT dept still needs more investigation. 
   - Also I have noticed some very weird behavior with outlook when using the extended scroll reports (freezing and locking up of outlook if scroll values sent by the keyboard are too large) need to test if this is only happening with hires scrolling enabled vs just extended scroll reports

Need to try and figure out what is going on with Linux as it should work just the same as the windows implementation (will need to try and get a USB Linux distro going to test to try and reproduce the reported lag on first report).

Finally I need to determine if these issues are surmountable with only device side code or if this feature will always be a little janky on Windows/Linux without an accompanying driver.

Still a lot to do on this but very open to getting help here from anyone who is willing (once I push the update so this is no longer so far behind the main develop branch).
